### PR TITLE
Remove Lucene from built-in recipes (Lombiq Technologies: OCORE-84)

### LINKED ISSUE #11203: Remove Lucene from built-in recipes
### Is your feature request related to a problem? Please describe.
Similar to https://github.com/OrchardCMS/OrchardCore/issues/11200 this is about making the built-in setup recipes lightweight: Lucene is a feature that you don't need on every site, especially that due to it storing files on the local file system that need to be preserved (i.e. not temp files) it needs special considerations for hosting (you need to take care of those files when moving the app between environments and doing deployments).

### Describe the solution you'd like
I propose to not enable Lucene by default in the following recipes:
- Headless: just remove it.
- Blank: just remove it.
- Blog: remove it and reimplement the `RecentBlogPosts` Query as an SQL Query. Since this recipe is also kind of an example of configuring Lucene, I wouldn't completely destroy this aspect; rather, the current configuration can remain in the theme as a separate, non-setup recipe that you can run if you want to do what the current recipe does, Query-wise (i.e. it would enable and configure Lucene and indexing, delete the `RecentBlogPosts` SQL Query and replace it with the current Lucene Query).

### Describe alternatives you've considered
The alternative is disabling the feature after every setup and re-implementing `RecentBlogPosts` as an SQL Query.

### PR DESCRIPTION
**Pull request by BanzragchUchral**
Fixes #11203

### CODE REVIEW

File: src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.setup.recipe.json [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+          "Description": "Anonymous role",
+          "Permissions": [
+            "ViewContent",
+            "QueryLuceneSearchIndex"
+          ]
+        }
+      ]
+    }
+  ]
+}
```
Comment (Skrypt) [inline src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.setup.recipe.json:78]: This means you need to add an extra line.

Comment (BanzragchUchral) [reply src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.setup.recipe.json]: Sorry, I didn't know.

File: test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs [hunk 2]
```diff
@@ -14,11 +16,15 @@ internal class PermissionContextAuthorizationHandler : AuthorizationHandler<Perm
         private readonly PermissionsContext _permissionsContext;
 
         // Used for http based graphql tests, marries a permission context to a request
-        public PermissionContextAuthorizationHandler(IHttpContextAccessor httpContextAccessor, IDictionary<string, PermissionsContext> permissionsContexts)
+        public PermissionContextAuthorizationHandler(
+            IServiceProvider serviceProvider,
+            IHttpContextAccessor mockHttpContextAccessor,
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:21]:
```suggestion
            IHttpContextAccessor httpContextAccessor,
```

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 3]
```diff
@@ -82,6 +104,70 @@ public virtual async Task InitializeAsync()
             GraphQLClient = new OrchardGraphQLClient(Client);
         }
 
+        public async Task RunLuceneRecipe(IShellHost shellHost)
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:107]:
```suggestion
        public async Task RunLuceneRecipeAsync(IShellHost shellHost)
```

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 4]
```diff
@@ -82,6 +104,70 @@ public virtual async Task InitializeAsync()
...
+            var shellScope = await shellHost.GetScopeAsync(TenantName);
+            await shellScope.UsingAsync(async scope =>
+            {
+                var shellFeaturesManager = scope.ServiceProvider.GetRequiredService<IShellFeaturesManager>();
+                var recipeHarvesters = scope.ServiceProvider.GetRequiredService<IEnumerable<IRecipeHarvester>>();
+
+                var recipeEventHandlers = new List<IRecipeEventHandler> { new RecipeEventHandler() };
+                var loggerMock = new Mock<ILogger<RecipeExecutor>>();
+
+                var recipeExecutor = new RecipeExecutor(
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:118]: Why not resolve `IRecipeExecutor` instead?

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 5]
```diff
@@ -1,13 +1,35 @@
...
+using OrchardCore.Recipes.Models;
+using OrchardCore.Recipes.Services;
+using OrchardCore.Settings;
 
 namespace OrchardCore.Tests.Apis.Context
 {
     public class SiteContext : IDisposable
     {
+        private const string luceneRecipePath = "Areas/TheBlogTheme/Recipes";
+        private const string luceneRecipeName = "blog.lucene.recipe.json";
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:31]: These are only related to the Blog recipes, so keep them in the Blog-related classes, not in this generic root one (`RunLuceneRecipe`'s name should also reflect this; it could accept a recipe name parameter to make it generally usable).

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 6]
```diff
@@ -82,6 +104,70 @@ public virtual async Task InitializeAsync()
...
+                    shellHost,
+                    scope.ShellContext.Settings,
+                    recipeEventHandlers,
+                    loggerMock.Object);
+
+                var features = await shellFeaturesManager.GetAvailableFeaturesAsync();
+                var recipes = await GetRecipesAsync(features, recipeHarvesters);
+
+                var recipe = recipes
+                    .FirstOrDefault(c => c.RecipeFileInfo.Name == luceneRecipeName && c.BasePath == luceneRecipePath);
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:128]: Use a proper variable name instead of `c`.

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 7]
```diff
@@ -82,6 +104,70 @@ public virtual async Task InitializeAsync()
...
+                var executionId = Guid.NewGuid().ToString("n");
+
+                await recipeExecutor.ExecuteAsync(
+                    executionId,
+                    recipe,
+                    new Dictionary<string, object>(),
+                    CancellationToken.None);
+            });
+
+            shellScope = await shellHost.GetScopeAsync(TenantName);
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:139]: Why run this again?

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 8]
```diff
@@ -82,6 +104,70 @@ public virtual async Task InitializeAsync()
...
+            await shellScope.UsingAsync(async scope =>
+            {
+                var luceneIndexSettingsService = new LuceneIndexSettingsService(
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:142]: Why not resolve this and `LuceneIndexingService` instead?

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 9]
```diff
@@ -109,6 +195,42 @@ public void Dispose()
...
+
+        private async Task<IEnumerable<RecipeDescriptor>> GetRecipesAsync(
+            IEnumerable<IFeatureInfo> features,
+            IEnumerable<IRecipeHarvester> recipeHarvesters)
+        {
+            var recipeCollections = await Task.WhenAll(recipeHarvesters.Select(x => x.HarvestRecipesAsync()));
+            var recipes = recipeCollections.SelectMany(x => x)
+                .Where(r => r.IsSetupRecipe == false &&
+                    !r.Tags.Contains("hidden", StringComparer.InvariantCultureIgnoreCase) &&
+                    features.Any(f => r.BasePath.Contains(f.Extension.SubPath, StringComparison.OrdinalIgnoreCase)));
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:207]: All this filtering is not necessary, since we're in a test, only the filter for the recipe name.

File: test/OrchardCore.Tests/Apis/Context/SiteContext.cs [hunk 10]
```diff
@@ -109,6 +195,42 @@ public void Dispose()
...
+
+            return recipes;
+        }
+
+        private class RecipeEventHandler : IRecipeEventHandler
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteContext.cs:212]: Why is this necessary?

File: test/OrchardCore.Tests/Apis/Context/SiteStartup.cs [hunk 11]
```diff
@@ -17,10 +18,15 @@ namespace OrchardCore.Tests.Apis.Context
     public class SiteStartup
     {
         public static ConcurrentDictionary<string, PermissionsContext> PermissionsContexts;
+        private static Mock<IHttpContextAccessor> mockHttpContextAccessor;
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/SiteStartup.cs:21]: Why does this need to be a static field instead of just constructing the mock in `ConfigureServices`?

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "Blog.Lucene",
-  "displayName": "Blog - Lucene",
-  "description": "Runs Lucene setup recipe and replaces RecentBlogPosts SQL query with Lucene.",
+  "name": "Blog.Lucene.Query",
```
Comment (Piedone) [inline src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json:2]: After running the recipe the query is empty:
image

File: src/docs/getting-started/starter-recipes.md [hunk 13]
```diff
@@ -82,20 +82,28 @@ TheBlogTheme is based on the [Start Bootstrap Clean Blog Theme](https://startboo
...
 - Bootstrap
 
-### Blog - Lucene Recipe
+### Blog - Lucene Query Recipe
 
-The Lucene recipe is an optional recipe in the TheBlogTheme. This recipe sets up
-Lucene indices and permissions. As an example, the recipe replaces the RecentBlogPosts
-SQL query with a Lucene query.
+The Lucene Query recipe is an optional recipe in the TheBlogTheme. 
+This recipe runs the Blog Search recipe and as an example,
```
Comment (Piedone) [inline src/docs/getting-started/starter-recipes.md:88]:
```suggestion
This recipe runs the Blog Lucene Search recipe and as an example,
```

File: test/OrchardCore.Tests/Apis/Context/BlogContext.cs [hunk 14]
```diff
@@ -1,14 +1,29 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Environment.Shell;
 
 namespace OrchardCore.Tests.Apis.Context
 {
     public class BlogContext : SiteContext
     {
+        private const string luceneRecipePath = "Areas/TheBlogTheme/Recipes";
+        private const string luceneRecipeName = "blog.lucene.query.recipe.json";
+        private const string luceneIndexName = "Search";
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/BlogContext.cs:11]: This is copied between the two contexts.

File: test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs [hunk 15]
```diff
@@ -14,11 +16,15 @@ internal class PermissionContextAuthorizationHandler : AuthorizationHandler<Perm
...
-        public PermissionContextAuthorizationHandler(IHttpContextAccessor httpContextAccessor, IDictionary<string, PermissionsContext> permissionsContexts)
+        public PermissionContextAuthorizationHandler(
+            IServiceProvider serviceProvider,
+            IHttpContextAccessor httpContextAccessor,
+            IDictionary<string, PermissionsContext> permissionsContexts)
         {
             _permissionsContext = new PermissionsContext();
 
-            var requestContext = httpContextAccessor.HttpContext.Request;
+            var httpContext = serviceProvider.GetRequiredService<IHttpContextAccessor>().HttpContext;
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:26]: This is also injected.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json (hunk 12 above)
Comment (BanzragchUchral) [reply src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json:2]: You need to reset the Search Lucene index because the content item was created before the Lucene indices.
image

File: test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs (hunk 15 above)
Comment (BanzragchUchral) [reply test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:26]: What do you mean by that? What should I do here?

File: src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json (hunk 12 above)
Comment (Piedone) [reply src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json:2]: This can be done from the recipe, so it's not a necessary manual step, no?

File: test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs (hunk 15 above)
Comment (BanzragchUchral) [reply test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:26]: That `IContextAccessor `is a mocked one. I only use it if this injected `httpContext `is null. I tried to use only the mocked `IContextAccessor`, but the GraphQLClient doesn't work with that.

Comment (Piedone) [reply test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:26]: When `PermissionContextAuthorizationHandler` is resolved (instantiated) the `IHttpContextAccessor` it gets injected into its ctor is the same that's in the injected `IServiceProvider` that you're resolving here (since that `IServiceProvider` is what `PermissionContextAuthorizationHandler` is also resolved from). So, there shouldn't be any difference, because you're accessing the same service from the same Di container.

Comment (BanzragchUchral) [reply test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:26]: I understand what you are saying, but the `IHttpContextAccessor` in the ctor is a mock object.

image

Comment (Piedone) [reply test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:26]: I see. `PermissionContextAuthorizationHandler` should use the same service in all cases. If you need to override what gets into the container, you can register the mock instance in `SiteStartup`.

File: src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json (hunk 12 above)
Comment (BanzragchUchral) [reply src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.lucene.query.recipe.json:2]: I look through the code, and the Lucene recipe step creates the index but doesn't process the old content items. I can modify the [LuceneIndexStep](https://github.com/OrchardCMS/OrchardCore/blob/339d74b251b53393dd510f8ad3027aa17d78500d/src/OrchardCore.Modules/OrchardCore.Lucene/Recipes/LuceneIndexStep.cs#L46) and make it possible, but I am not sure if there is any reason we haven't done it before.

File: test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs [hunk 16]
```diff
@@ -17,14 +15,15 @@ internal class PermissionContextAuthorizationHandler : AuthorizationHandler<Perm
...
-            IServiceProvider serviceProvider,
             IHttpContextAccessor httpContextAccessor,
             IDictionary<string, PermissionsContext> permissionsContexts)
         {
             _permissionsContext = new PermissionsContext();
 
-            var httpContext = serviceProvider.GetRequiredService<IHttpContextAccessor>().HttpContext;
-            var requestContext = httpContext != null ? httpContext.Request : httpContextAccessor.HttpContext.Request;
+            if (httpContextAccessor.HttpContext == null)
+                return;
```
Comment (Piedone) [inline test/OrchardCore.Tests/Apis/Context/AuthenticationContext.cs:22]: In OC this is the convention:

```suggestion
            if (httpContextAccessor.HttpContext == null)
            {
                return;
            }
```

**Review by Piedone — CHANGES_REQUESTED** [review]

**Comment by Skrypt** [comment]
I will do tomorrow.

**Comment by Piedone** [comment]
Thanks!

**Review by Piedone — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Lucene/Recipes/LuceneIndexStep.cs [hunk 17]
```diff
@@ -44,6 +44,7 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
                     {
                         luceneIndexSettings.Value.IndexName = luceneIndexSettings.Key;
                         await _luceneIndexingService.CreateIndexAsync(luceneIndexSettings.Value);
+                        await _luceneIndexingService.ProcessContentItemsAsync(luceneIndexSettings.Key);
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Lucene/Recipes/LuceneIndexStep.cs:47]: This can cause a request timeout which is why it was not added. In this case, if the recipe fails to execute then executing it a second time would probably not work.

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Lucene/Recipes/LuceneIndexStep.cs:47]: It should be a fire and forget type of execution here. Start the indexing background task for example.

**Review by Skrypt — CHANGES_REQUESTED** [review]
Looks good except the recipe step should not process the content items. Also, it may fail to index these Blog Posts since they are probably not added yet to the IndexingTask table.

**Comment by BanzragchUchral** [comment]
You mean before running the new Lucene recipe, right? I can see a few records here. I am not sure if this is what you wanted to see.

image

**Comment by Skrypt** [comment]
Ok, this works with TheBlog recipe because we enable the Lucene feature from the setup. But let's say we do the same but with the SaaS recipe then. I'm pretty sure that the IndexingTask table will not be there. So, if we create content items before enabling the Lucene feature then these content items will not be added to the IndexingTask table.

I'm ok to merge the current PR as is since this is a known issue that needs to be fixed. But we could fix it with a different PR.

**Review by Skrypt — APPROVED** [review]

**Review by Piedone — APPROVED** [review]
